UnitWorld_shared/shared/communication/MessageWrapper.cpp

```diff
@@ -0,0 +1,86 @@
+#include "MessageWrapper.h"
+
+#include "CompleteGameStateMessage.h"
+#include "InvalidMessage.h"
+
+#include "MessageType.h"
+
+#include <nlohmann/json.hpp>
+
+#include <chrono>
+
+namespace uw
+{
+    NLOHMANN_JSON_SERIALIZE_ENUM(MessageType, {
+        {InvalidMessageType, nullptr},
+        {CompleteGameStateMessageType, "complete-game-state"},
+    });
+
+    const std::string MessageWrapper::MESSAGE_TYPE_JSON_ATTRIBUTE = "type";
+    const std::string MessageWrapper::MESSAGE_DATA_JSON_ATTRIBUTE = "data";
+    const std::string MessageWrapper::MESSAGE_TIMESTAMP_JSON_ATTRIBUTE = "timestamp";
+}
+
+using namespace uw;
+
+MessageWrapper::MessageWrapper(const std::string& json) :
+    _timestamp(jsonToTimestamp(json)),
+    _innerMessage(jsonToMessage(json)),
+    _json(json)
+{}
+
+MessageWrapper::MessageWrapper(const std::shared_ptr<const Message> message) :
+    _timestamp(std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::time_point_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now()).time_since_epoch()).count()),
+    _innerMessage(message),
+    _json(wrapMessageToJson(message, _timestamp))
+{}
+
+std::shared_ptr<Message const> MessageWrapper::innerMessage() const
+{
+    return _innerMessage;
+}
+
+std::string MessageWrapper::json() const
+{
+    return _json;
+}
+
+unsigned long long MessageWrapper::timestamp() const
+{
+    return _timestamp;
+}
+
+std::string MessageWrapper::wrapMessageToJson(const std::shared_ptr<const Message> message, const unsigned long long& timestamp)
+{
+    nlohmann::json jsonMessage = {
+        {MESSAGE_TYPE_JSON_ATTRIBUTE, message->messageType()},
+        {MESSAGE_DATA_JSON_ATTRIBUTE, nlohmann::json::parse(message->toJsonData())},
+        {MESSAGE_TIMESTAMP_JSON_ATTRIBUTE, timestamp}
+    };
+
+    return jsonMessage.dump();
+}
+
+std::shared_ptr<const Message> MessageWrapper::jsonToMessage(const std::string& json)
+{
+    nlohmann::json parsedJson = nlohmann::json::parse(json);
+    const auto messageType = parsedJson.at(MESSAGE_TYPE_JSON_ATTRIBUTE).get<MessageType>();
+    const auto messageData = parsedJson.at(MESSAGE_DATA_JSON_ATTRIBUTE).dump();
+
+    if (messageType == MessageType::CompleteGameStateMessageType)
```

Comment:
Tell, don't ask.

Also, I see usages of `nlohmann::json` at too many places. You might want to abstract this dependency away.

Reply:
what do you mean tell don't ask? Should I ask the message classes statically if the message type belongs to them?

What if I have a map of message type, mapping to a function that parses the message? this method would provide O(1) finding the message type but I can't see how I'd implement that without checking myself the message type

Reply:
Well this should technically be in the "communication" namespace, where we transfer information between the server and the client. I did do an effort in making the json dependency local to that, do you mean I should wrap the dependency further? Json provides very useful and hard to forward STL shortcuts, and I'm pretty sure I don't want to interface all of them

Reply:
> what do you mean tell don't ask?

By that, I mean that you ask `nlohmann::json` information about its inner state, and then you act according to what you got (that results in an `if` statement). What if you want to add more `MessageType`? You probably end up having to change this function too (according to what I understand from your design). So you might want to *tell* `nlohmann::json parsedJson` to wrap its underlying value to an instance of the `Message` class. And, since I think you can't change yourself this library (i.e. add functions), then you might also want to abstract it away, so that you gain control over your code.

Next, I think this idea does not prevent the map implementation you describe. You might, for instance, code this inside the interface that would replace `nlohmann::json parsedJson`.

Reply:
As I see it, would wrapping the json dependency simply push the "tell don't ask" further down the chain? I'd still have a chain of if checking what message type it really was.

Solving the tell don't ask here means to defer the message selection to the messages themselves, or to a "message selector" so to say, which is what I'm doing here in the wrapper am I not?

Reply:
> As I see it, would wrapping the json dependency simply push the "tell don't ask" further down the chain? I'd still have a chain of if checking what message type it really was.

You are partially right. If you abstract away the json dependency *alone*, then you end up deferring the `if..else if... else...` to some other class, and it this class is still not shielded against change.

I thought a little bit more about this in depth, and, in fact, the the "message selector" you are talking about really is the responsibility of the `MessageType` class. I might be wrong, but accordingly to my current understanding, you should in fact defer the message selection mechanism to the `MessageType` class.

Reply:
> Well this should technically be in the "communication" namespace, where we transfer information between the server and the client. I did do an effort in making the json dependency local to that, do you mean I should wrap the dependency further? Json provides very useful and hard to forward STL shortcuts, and I'm pretty sure I don't want to interface all of them

That is an issue I didn't think about, so I guess you *can* leave the explicit dependency in your code.

Nevertheless, at least I think you should replace those `if...else` statements by adding polymorphic members to `MessageType`.

Reply:
You can't add methods in enums in cpp, so instead of having the message type selection in message type, I put them in a "message selection" class, which is called messageWrapper

Reply:
Why not using a class instead of an enum then? You could imitate the idea of an enum by using a private constructor, and implement members using composition (so that only one `MessageType` class need to exist). I bet there would be no more need for this `MessageWrapper` class.